src/main/java/com/fasterxml/jackson/core/util/BufferRecyclerPool.java

```diff
@@ -1,17 +1,157 @@
+        THREAD_LOCAL(new ThreadLocalRecyclerPool()), 
+        CONCURRENT_DEQUEUE(new ConcurrentDequePool()),
+        LOCK_FREE(new LockFreePool());
+
+        private final BufferRecyclerPool pool;
+
+        PoolStrategy(BufferRecyclerPool pool) {
+            this.pool = pool;
+        }
+
+        public BufferRecyclerPool getPool() {
+            return pool;
+        }
+    }
+
+    /**
+     * Default {@link BufferRecyclerPool} implementation that uses
+     * {@link ThreadLocal} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ThreadLocalRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return BufferRecyclers.getBufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that does not use
+     * any pool but simply creates new instances when necessary.
+     *
+     * @since 2.16
+     */
+    class NonRecyclingRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * {@link ConcurrentLinkedDeque} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ConcurrentDequePool implements BufferRecyclerPool {
+        private final Deque<BufferRecycler> pool = new ConcurrentLinkedDeque<>();
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return getBufferRecycler().withPool(this);
+        }
+
+        private BufferRecycler getBufferRecycler() {
+            BufferRecycler bufferRecycler = pool.pollFirst();
+            return bufferRecycler != null ? bufferRecycler : new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler bufferRecycler) {
+            pool.offerLast(bufferRecycler);
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * a lock free linked list for recycling instances.
+     *
+     * @since 2.16
+     */
+    class LockFreePool implements BufferRecyclerPool {
+        private final AtomicReference<LockFreePool.Node<BufferRecycler>> head = new AtomicReference<>();
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return getBufferRecycler().withPool(this);
+        }
+
+        private BufferRecycler getBufferRecycler() {
+            for (int i = 0; i < 3; i++) {
```

Comment:
Could use some commentary on why looping for 3 times etc.

Reply:
Done.